### DISCUSSION TITLE
Add Conditional Expression `When()` to run query builder when condition is true

### DISCUSSION THREAD

**Discussion by gmhafiz** [ent/ent#2370] [Ideas]
When building a query, there are cases when you want to add further query builder(s) if and only if certain conditions are met. In sql, we can use `CASE` operator. But this solution generalize to other database systems. 

For an example, currently if we are deciding whether to paginate or not, we check the condition is true in the business logic:

```go
paginate := <some business logic>
limit := 10
offset := 0

q := client.User.Query().
    Where(wherePredicate)
	
if paginate {
    q = q.
	    Limit(limit).
	    Offset(offset).
	    Order(ent.Asc(field))
}
	
res, err = q.All(ctx)
```
While this is clear what it is doing, I feel it can be improved. For example:

1. Due to indentation, all 4 `q` variables are now on four column levels, reducing readability because the variables are not in a single happy path https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88
2. Three separate blocks of code, (`q`, condition check, execute)

Using a ` When()` query builder will turn the code into

```go
limit := 10
offset := 0
res, err := client.User.Query().
    Where(wherePredicate).
    When(paginate, func(q *ent.userQuery) {
	q.
	    Limit(limit).
	    Offset(offset).
	    Order(ent.Asc(field))
    }).
    All(ctx)
```

In my opinion, this looks more elegant for the reason:
1. Only two column levels for `q`.

As far as i know, other ORMs that have a `When()` builder are

1. [Laravel](https://laravel.com/docs/9.x/queries#conditional-clauses)
3. [Django](https://docs.djangoproject.com/en/4.0/ref/models/conditional-expressions/#when)

Other popular ORMs that do not have it AFAIK, are
1. Sqlalchemy
2. Rails ActiveRecord
4. Sqlboiler
5. Gorm
6. Rust's [Diesel](https://github.com/diesel-rs/diesel/issues/455)
7. TypeOrm
	
Implementing it is easy. Only 7 additional lines are required to `query.tmpl`. It takes 7 out of 589 lines in `user_query.go`.

```go
// When runs the provided builder(s) if and only if condition is true.
func ({{ $receiver }} *{{ $builder }}) When(condition bool, action func(builder *{{ $builder }})) *{{ $builder }} {
	if condition {
		action({{ $receiver }})
	}

	return {{ $receiver }}
}
```

I also have a test for it, in which you can review a pull request if this new functionality is desirable.

**Comment by gmhafiz** [comment]
Related variation: https://github.com/ent/ent/pull/2269/commits/0d4bb397299ee42176c0ee4a661c7809a9f990c0

**Reply by febxr** [reply]
good, wanted. it's elegant and cool.

**Comment by a8m** [comment]
Hey @gmhafiz! Adding here my comment from our Discord:

> Maybe adding a generic API under the `dialect/sql` package is simpler?
> ```go
> func When[P ~func(*sql.Selector), Q interface{ Where(...P) Q }](q Q, cond bool, p P) Q {
>    if cond {
>        q.Where(p)
>    }
>    return q
> }
> ```
> Usage:
> ```go
> // Create a new Query.
> q := When(client.File.Query(), cond, file.Name("etc"))
> 
> // Apply on existing Query.
> q = When(q, cond, file.Name("etc"))
> ```
